src/shaders/glsl/mesh.fragment.shaded.js

```diff
@@ -16,7 +18,12 @@ vec4 getShadedColor(vec4 rgba) {
   vec3 light = normalize(vLight);
   vec3 position = normalize(vPosition);
   
-  float side    = gl_FrontFacing ? -1.0 : 1.0;
+  // Workaround to avoid gl_FrontFacing. See https://github.com/unconed/mathbox/pull/26
+  vec3 pdx = dFdx(vPosition);
+  vec3 pdy = dFdy(vPosition);
+  bool frontFacting = dot(vNormal, cross(pdx, pdy)) > 0.0;
```

Comment:
whoops, slight typo with `frontFacting` vs `frontFacing`.

Wow, there is some magic going on in glsl. SO, indulge me in trying to understand what this does:

- `dfDx` and `dfDy`  must return vectors; are they giving the derivative of the x and y components of `vPosition`?
- If that's right, the cross product will always be orthogonal... so we care if the magnitude is positive or negative.
- and then `dot(...) > 0` is telling us, "is vNormal pointing in the same direction"

But I am missing my concepts here, since I would have expected `vNormal` to be in reference to some plane... ah!!

I would love a one sentence description of the algorithm here for my own edification, and probably for future folks trying to debug.

Reply:
To be honest, I am not 100% clear on the coordinate systems involved here. I *think* the following sentence would be accurate:

"vNormal (calculated in the vertex shader) and cross(pdx, pdy) (calculated in this fragment shader) are both surface normals, but might be in opposite directions depending on which side of the surface the camera is, hence dot product".

My impression is that `vNormal` is calculated differently for the different objects, that's why there's `surface.position.normal`, `face.position.normal`, and `strip.position.normal`. [(but represents a surface normal in all cases), but represents a surface normal in all three cases.] 
If you look at how `vNormal` in `surface.position.normal` is calculated, you'll see its also basically a cross product of derivatives. But I suspect the coordinates are slightly different (potentially reversed) hence the two cross products may differ in direction. (Experimentally, they definitely **do** differ in direction, since both conditions in the `frontFacing ? 1.0 : -1.0` are getting triggered.)

Understanding the shader stuff better is definitely a goal :/ I'm hoping not to tinker with it too much.

In terms of posterity... I decided just leave a link to the PR as the code comment. (Though git history does that, too).

Reply:
In particular, one thing I am confused about is `getPosition` in the shaders. A lot of the vertex shaders include
```
// External
vec3 getPosition(vec4 xyzw, float canonical);
```

and then `getPosition` is used. But I can't find where the heck this "External" function is coming from...